src/graphics/shader-node-registry.js

```diff
@@ -0,0 +1,88 @@
+import { ShaderGraphNode } from '../scene/materials/shader-graph-node.js';
+import { hashCode } from '../core/hash.js';
+/**
+ * @private
+ * @class
+ * @name pc.ShaderNodeRegistry
+ * @classdesc Container for all Shader Graph Core Nodes available to this application.
+ * @description Create an instance of a pc.ShaderNodeRegistry.
+ * @param {pc.Application} app - Application to attach registry to.
+ */
+function ShaderNodeRegistry(app) {
+    this.app = app;
+    this._nodeDef = { };
+    this._nodeCache = { };
+}
+
+ShaderNodeRegistry.prototype.constructor = ShaderNodeRegistry;
+
+ShaderNodeRegistry.prototype.destroy = function () {
+    this.app = null;
+};
+
+/**
+ * @private
+ * @function
+ * @name pc.ShaderNodeRegistry#register
+ * @description Register core nodes with registry.
+ * @param {object|string} coreNodeList - core node list object or JSON string
+ */
+ShaderNodeRegistry.prototype.register = function (coreNodeList) {
+    if (coreNodeList) {
+        if (typeof coreNodeList === 'string') {
+            coreNodeList = JSON.parse(coreNodeList);
+        }
+        var self = this;
+        Object.keys(coreNodeList).forEach(function (key) {
+            self._nodeDef[key] = coreNodeList[key];
+            self._nodeCache[key] = {};
+        });
+    }
+};
+
+ShaderNodeRegistry.prototype._genVariantKey = function (argTypes, options) {
+    var variantKey = 'varKey';
+
+    for (var argIndex = 0; argIndex < argTypes.length; argIndex++) {
+        variantKey += '_' + argTypes[argIndex];
+    }
+
+    if (options) {
+        var optionKeys = Object.keys(options);
+        for (var keyIndex = 0; keyIndex < optionKeys.length; keyIndex++) {
+            variantKey += '_' + optionKeys + '_' + options[optionKeys[keyIndex]];
+        }
+    }
+
+    return hashCode(variantKey);
+};
+
+ShaderNodeRegistry.prototype.get = function (name, argTypes, options) {
+    if (this._nodeCache[name]) {
```

Comment:
you should store the result of this instead of executing it many times, as each of these is a dictionary lookup.
`this._nodeCache[name]`

Reply:
Not sure what you mean here - the node registry is just a single instance per app - and this core node cache is only used when building graphs - which happens only on shader graph editing/building or generation.

Reply:
What I meant what that instead of calling this many times:
`this._nodeCache[name]`
ideally you store it in a var and reuse:
`var namedNode = this._nodeCache[name];`

Reply:
ah - okay - good point - and done.